File: mrt.hs
import System.IO
import Data.Maybe


pl = [2,3,5,7,11,13,17,19,23]

mrt 2 = True
mrt x = odd x && all (\a -> work s x (expMod a d x) ) [2..(x-2)] where
  (s,d) = sd (x-1)

fastMRT 2 = True
fastMRT x 
  | x > 23 =odd x && all (\a -> work s x (expMod a d x) ) pl 
  | x <= 23 = odd x && all (\a -> work s x (expMod a d x) ) [2..(x-2)]
  where
  (s,d) = sd (x-1)

sd n = if odd n then (0,n) else (\(x,y) -> (x+1,y)) (sd (n `div` 2))

-- work s n x
work 0 n x = (x == 1) || (x == n-1)
work s n x
  |(f == 1) && (x /= 1) && (x /= n-1) = False
  |otherwise = work (s-1) n f  where f = (x*x) `mod` n


expMod a 0 m = 1
expMod a d m
  | even d    = expMod ((a * a) `mod` m) (d `div` 2) m
  | otherwise = (a * expMod a (d - 1) m) `mod` m

firstWitness :: Integer -> Maybe Integer
firstWitness x = go x pl where
  go :: Integer -> [Integer] -> Maybe Integer
  go n [] = Nothing
  go n (a:as) = if work s n (expMod a d n) then go n as else Just a where (s,d) = sd (n-1)


cleanData :: [Maybe a] -> [a]
cleanData [] = []
cleanData (x:xs) = case x of
  Nothing -> cleanData xs
  Just t -> t : cleanData xs


substitute :: (Eq a) => [a] -> [(a, a)] -> [a]
substitute xs subs = map (\x -> fromMaybe x (lookup x subs)) xs

frequencyTable :: [Integer] -> Integer -> [(Integer, Int)]
frequencyTable xs n = [(i, count i) | i <- [1..n]]
  where
    count x = length (filter (== x) xs)


main = do
  let n = 2^21
  let clean_data = cleanData $ map firstWitness [25,27..n]
  let sub_data = substitute clean_data (zip pl [1..])
  let table = frequencyTable sub_data 9
  writeFile "dataH.txt" (unlines $ map (\(x,y) -> show x ++ ":" ++ show y) table)

File: mrt.py
## python translation of the haskell code made for no purpose but fun (and speed tests)

import math
from typing import List, Optional, Tuple

pl = [2, 3, 5, 7, 11, 13, 17, 19, 23]

def mrt(x: int) -> bool:
    if x == 2:
        return True
    s, d = sd(x - 1)
    return x % 2 == 1 and all(work(s, x, exp_mod(a, d, x)) for a in range(2, x - 1))

def fast_mrt(x: int) -> bool:
    if x == 2:
        return True
    s, d = sd(x - 1)
    if x > 23:
        return x % 2 == 1 and all(work(s, x, exp_mod(a, d, x)) for a in pl)
    else:
        return x % 2 == 1 and all(work(s, x, exp_mod(a, d, x)) for a in range(2, x - 1))

def sd(n: int) -> Tuple[int, int]:
    if n % 2 == 1:
        return (0, n)
    else:
        s, d = sd(n // 2)
        return (s + 1, d)

def work(s: int, n: int, x: int) -> bool:
    if s == 0:
        return x == 1 or x == n - 1
    f = (x * x) % n
    if f == 1 and x != 1 and x != n - 1:
        return False
    return work(s - 1, n, f)

def exp_mod(a: int, d: int, m: int) -> int:
    if d == 0:
        return 1
    if d % 2 == 0:
        return exp_mod((a * a) % m, d // 2, m)
    else:
        return (a * exp_mod(a, d - 1, m)) % m

def first_witness(x: int) -> Optional[int]:
    return go(x, pl)

def go(n: int, a_list: List[int]) -> Optional[int]:
    if not a_list:
        return None
    a, *as_ = a_list
    s, d = sd(n - 1)
    if work(s, n, exp_mod(a, d, n)):
        return go(n, as_)
    else:
        return a

def clean_data(data: List[Optional[int]]) -> List[int]:
    return [t for x in data if x is not None for t in (x,)]

def substitute(xs: List[int], subs: List[Tuple[int, int]]) -> List[int]:
    subs_dict = dict(subs)
    return [subs_dict.get(x, x) for x in xs]

def frequency_table(xs: List[int], n: int) -> List[Tuple[int, int]]:
    def count(x: int) -> int:
        return sum(1 for i in xs if i == x)
    return [(i, count(i)) for i in range(1, n + 1)]

if __name__ == "__main__":
    n = 2**25
    clean_data_result = clean_data(map(first_witness, range(25, n + 1, 2)))
    sub_data = substitute(clean_data_result, zip(pl, range(1, len(pl) + 1)))
    table = frequency_table(sub_data, 9)
    
    with open("data.txt", "w") as f:
        f.write("\n".join(f"{x}:{y}" for x, y in table))